### DISCUSSION TITLE
Updating translatable fields to null or empty is not working as expected

### DISCUSSION THREAD

**Discussion by kskrlinnorth2** [spatie/laravel-translatable#356] [Q&A]
Example for create:
```
Model::create([
    'name' => ['en' => 'Name en', 'hr' => 'Name hr'],
    'description' => ['en' => 'Desc en', 'hr' => 'Desc hr'],
]);
```
In DB I get data saved as JSON (as expected).

Example for update:
```
$model->update([
    'name' => ['en' => 'Name en', 'hr' => 'Name hr'],
    'description' => null,
]);
```
In DB I am getting `{"en":null}` instead of null. Why isn't saved as null (it is nullable column in DB), but as `en: null`?

**Comment by anibalealvarezs** [comment]
Unless you include a key/value pair specifying which language the `null` value refers to, it will assign it to the current app language.
Since your current language is `en`, it will be stored as `{"en": null}`

Remember that the mutation will work for each field independently. So, it will work the same as

```php
$model->update([
    'name' => ['en' => 'Name en', 'hr' => 'Name hr'],
]);
$model->update([
    'description' => null,
]);
```

The `update()` method you're using is related to the Model, so the mutation will work for all the fields you have specified in the model definition. I assume you're using the `find()` method from the Builder to get the model instance: https://laravel.com/api/9.x/Illuminate/Database/Eloquent/Builder.html#method_find

A different case would have been using the Collection `update()` method, which would have ignored the mutation and you would have been able to assign a custom value to the field.

For example:

```php
$collection = Model::get() // <= This is a "collection", not a "model instance"
$collection->update([
    'description' => null,
])
```

That's because the `get()` method returns a collection instead of a model instance: https://laravel.com/api/9.x/Illuminate/Database/Eloquent/Builder.html#method_get

**Comment by kratos-digital** [comment]
I also stumbled on this problem today after client reported that current translations aren't getting removed when values for each locale is empty on update.
I fixed it by overriding the default HasTranslation trait which works for my case.

```
<?php

namespace App\Http\Traits;

use Spatie\Translatable\HasTranslations as BaseHasTranslations;

trait HasTranslations
{
    use BaseHasTranslations;
    
    public function setAttribute($key, $value)
    {
        // Force prop to be set as null
        if ($value === null) {
            $this->attributes[$key] = null;
            return;
        }

        if ($this->isTranslatableAttribute($key) && is_array($value)) {
            return $this->setTranslations($key, $value);
        }

        // Pass arrays and untranslatable attributes to the parent method.
        if (! $this->isTranslatableAttribute($key) || is_array($value)) {
            return parent::setAttribute($key, $value);
        }

        // If the attribute is translatable and not already translated, set a
        // translation for the current app locale.
        return $this->setTranslation($key, $this->getLocale(), $value);
    }
}
```

**Comment by aldesrahim** [comment]
+1 @kratos-digital 

Based on the code in [HasTranslations](https://github.com/spatie/laravel-translatable/blob/cd7dc7c89cf74cda2b687c989cc57b2df1ebecc4/src/HasTranslations.php#L58-L69)

You could override the `setAttribute` method and update the first IF condition to this:
```if (!$this->isTranslatableAttribute($key) || is_null($value))```

It would look something like this:
```
public function setAttribute($key, $value)
{
    if (!$this->isTranslatableAttribute($key) || is_null($value)) {
        return parent::setAttribute($key, $value);
    }

    if (is_array($value) && (!array_is_list($value) || count($value) === 0)) {
        return $this->setTranslations($key, $value);
    }

    return $this->setTranslation($key, $this->getLocale(), $value);
}
```
